### ISSUE TITLE
Slow initialisation: four second delay to load the opensc PKCS11 module

### ISSUE THREAD

**Issue by minfrin** [OpenSC/OpenSC#2444]
### Problem Description

When adding opensc to a command line tool using p11-kit to manage smartcards, the `p11_kit_modules_load_and_initialize` function takes approximately four seconds to return.

This drastically slows down the "command completion" feature of the command line tool.

During the four second pause, communication occurs to the smartcard, even though all we're doing at this point is initialising the module.

Removing the smartcard from the reader causes the pause to go away.

### Proposed Resolution

Identify what's taking four seconds and remove it.

### Steps to reproduce

MacOS Big Sur. pk11-kit and opensc v0.21.0 provided by Darwinports.

Use Redwax Tool from https://redwax.eu/rt/#getting-involved

Card reader from ACS or Feitian, MyEID card.

### Logs

Tool running with four second delay:

```
Little-Net:redwax-tool minfrin$ date; ./redwax-tool --pkcs11-in pkcs11:; date
Mon Nov 15 18:25:49 SAST 2021
pkcs11-in: root: CN=GlobalSign Root CA,OU=Root CA,O=GlobalSign nv-sa,C=BE
pkcs11-in: root: CN=GlobalSign,O=GlobalSign,OU=GlobalSign Root CA - R2
[snip many CA certs]
pkcs11-in: root: CN=ANF Secure Server Root CA,OU=ANF CA Raiz,O=ANF Autoridad de Certificacion,C=ES,serialNumber=G63287510
pkcs11-in: root: CN=Certum EC-384 CA,OU=Certum Certification Authority,O=Asseco Data Systems S.A.,C=PL
pkcs11-in: root: CN=Certum Trusted Root CA,OU=Certum Certification Authority,O=Asseco Data Systems S.A.,C=PL
pkcs11-in: certificate: xxx
pkcs11-in: intermediate: CN=GlobalSign PersonalSign 2 CA - SHA256 - G3,O=GlobalSign nv-sa,C=BE
pkcs11-in: certificate: xxx
Mon Nov 15 18:25:53 SAST 2021
Little-Net:redwax-tool minfrin$ 
```

**Comment by frankmorgner** [comment]
try running `date; OPENSC_DEBUG=20 ./redwax-tool --pkcs11-in pkcs11:; date` to get a debug log. the included timestamps should show where the delay is located.

**Comment by Jakuje** [comment]
You can try to enable file caching, which can speed up the card detection and the whole response time for p11-kit. Try to add `use_file_caching = true;` under the `framework pkcs15` block in `opensc.conf`.

**Comment by minfrin** [comment]
Looking at the debug output, there does not seem to be a delay in any single one step, what there does seem to be is a large amount of traffic is sent backwards and forwards to/from the card.

Using a simplified version of the command line that just runs the initialise step and nothing else, we have this:

```
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 15:34:46 SAST 2021
Tue Nov 16 15:34:50 SAST 2021
```

We then do this, and look at the results:

```
date; OPENSC_DEBUG=20 ./redwax-tool 2> /tmp/out; date
```

The first part of the log opens the smartcard reader with PCSC, and takes 61ms:

```
P:58732; T:0x4748676608 15:30:04.096 [opensc-pkcs11] ctx.c:724:process_config_file: Used configuration file '/opt/local/etc/opensc.conf'
[snip]
P:58732; T:0x4748676608 15:30:04.157 [opensc-pkcs11] reader-pcsc.c:637:pcsc_connect: Initial protocol: T=1
```

Now we start trying ATRs, this takes 19ms:

```
P:58732; T:0x4748676608 15:30:04.157 [opensc-pkcs11] card.c:1199:match_atr_table: ATR try : 3B:FE:18:00:00:81:31:FE:45:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:FA
P:58732; T:0x4748676608 15:30:04.157 [opensc-pkcs11] card.c:1199:match_atr_table: ATR try : 3B:8E:80:01:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:18
[snip]
P:58732; T:0x4748676608 15:30:04.176 [opensc-pkcs11] card.c:1199:match_atr_table: ATR try : 3B:8B:01:52:75:74:6F:6B:65:6E:6C:69:74:65:C2
P:58732; T:0x4748676608 15:30:04.176 [opensc-pkcs11] card.c:1199:match_atr_table: ATR try : 3B:9E:96:00:52:75:74:6F:6B:65:6E:4C:69:74:65:53:43:32
```

At this point our card is discovered, and we start sending a total of 68 APDUs (and receiving responses) over the next 3811ms:

```
Little-Net:redwax-tool minfrin$ cat /tmp/out | grep "reader-pcsc.c:323:pcsc_transmit"
P:58732; T:0x4748676608 15:30:04.177 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:04.246 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:04.276 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:04.336 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
[snip]
P:58732; T:0x4748676608 15:30:07.830 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:07.882 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:07.935 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:07.988 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
```

Many of the responses contain blank blocks of data, this is one example:

```
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] apdu.c:546:sc_transmit_apdu: called
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] card.c:473:sc_lock: called
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] apdu.c:513:sc_transmit: called
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] apdu.c:363:sc_single_transmit: called
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] apdu.c:370:sc_single_transmit: CLA:0, INS:B0, P1:1, P2:FE, data(0) 0x0
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] reader-pcsc.c:324:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 01 FE FF .<B0>.<FE><FF>
P:58732; T:0x4748676608 15:30:05.005 [opensc-pkcs11] reader-pcsc.c:242:pcsc_internal_transmit: called
P:58732; T:0x4748676608 15:30:05.058 [opensc-pkcs11] reader-pcsc.c:333:pcsc_transmit: 
Incoming APDU (257 bytes):
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 90 ................
00                                              .
P:58732; T:0x4748676608 15:30:05.058 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:58732; T:0x4748676608 15:30:05.058 [opensc-pkcs11] apdu.c:535:sc_transmit: returning with: 0 (Success)
```

The actual APDUs seem to be taking ages to transmit, in this example it takes 48ms to send 7 bytes and receive 41 bytes in return:

```
P:58732; T:0x4748676608 15:30:04.392 [opensc-pkcs11] apdu.c:370:sc_single_transmit: CLA:0, INS:A4, P1:8, P2:0, data(2) 0x7ffee4553392
P:58732; T:0x4748676608 15:30:04.392 [opensc-pkcs11] reader-pcsc.c:323:pcsc_transmit: reader 'ACS ACR 38U-CCID'
P:58732; T:0x4748676608 15:30:04.392 [opensc-pkcs11] reader-pcsc.c:324:pcsc_transmit: 
Outgoing APDU (7 bytes):
00 A4 08 00 02 50 15 .....P.
P:58732; T:0x4748676608 15:30:04.392 [opensc-pkcs11] reader-pcsc.c:242:pcsc_internal_transmit: called
P:58732; T:0x4748676608 15:30:04.440 [opensc-pkcs11] reader-pcsc.c:333:pcsc_transmit: 
Incoming APDU (41 bytes):
6F 25 81 02 7F FF 82 01 38 83 02 50 15 86 03 11 o%......8..P....
1F FF 85 02 00 02 8A 01 07 84 0C A0 00 00 00 63 ...............c
50 4B 43 53 2D 31 35 90 00                      PKCS-15..
P:58732; T:0x4748676608 15:30:04.440 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
```

To summarise:

- In the initialisation step, we are doing lots of non-init stuff, like downloading all data on the smartcard.
- The data transfer seems to be very very slow.
- We seem to be requesting a lot of data containing just zeros.

**Comment by minfrin** [comment]
While file caching speeds things up, end users will have to change this default to make this work (first slow below, rest sped up).

```
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 16:24:13 SAST 2021
Tue Nov 16 16:24:17 SAST 2021
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 16:24:20 SAST 2021
Tue Nov 16 16:24:21 SAST 2021
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 16:24:22 SAST 2021
Tue Nov 16 16:24:22 SAST 2021
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 16:24:23 SAST 2021
Tue Nov 16 16:24:24 SAST 2021
Little-Net:redwax-tool minfrin$ date; ./redwax-tool; date
Tue Nov 16 16:24:25 SAST 2021
Tue Nov 16 16:24:25 SAST 2021
```

The above doesn't explain why the data transfer is so slow.

**Comment by frankmorgner** [comment]
contact-based communication is quite slow. contact-less can be a lot better. way better is an integrated solution e.g. yubikey, which has the power to use the USB 2.0 maximum speed. For compensation of the hardware limitations, you could look into fine tuning the card driver and PKCS#15 profile:
- typically, files are read from beginning to end, even though only some data may have been used in a fixed size file. Use the ASN.1 data to determine the actual length of usable data
- use extended length APDUs (and readers) to get higher throughput
- go through the log and ask yourself whether a single command was needed or not and try to avoid card commands if possible (file cache is a low hanging fruit)

I once did this kind of optimization for sc-hsm and reduced the time for card detection from 1.4 seconds to 0.2 seconds. 1 second improvement doesn't sound like a lot, but in case of windows logon this initialization is done for around 10 different processes. In total this improved windows logon by 10-15 seconds, which was astonishing.

**Comment by frankmorgner** [comment]
> While file caching speeds things up, end users will have to change this default to make this work (first slow below, rest sped up).

You have my vote when it comes to changing the file caching to be switched on by default. However, this may confuse some users, because for years the default was to not store anything on disk since. Storing certificates on disk may leak personal information, for example.

**Comment by Jakuje** [comment]
> > While file caching speeds things up, end users will have to change this default to make this work (first slow below, rest sped up).
> 
> You have my vote when it comes to changing the file caching to be switched on by default. However, this may confuse some users, because for years the default was to not store anything on disk since. Storing certificates on disk may leak personal information, for example.

FYI we ship a configuration which uses file caching by default in Fedora and RHEL for some time. I think with visible release notes, this is fine if you want to change it in the next release too. One can expect that something from the card might be logged when you plug the card in some reader (or approach a contactless one.
